src/Development/IDE/Core/Rules.hs

```diff
@@ -867,21 +852,39 @@ regenerateHiFile sess f = do
             -- on the parsed module and the typecheck rules
             (diags', tmr) <- typeCheckRuleDefinition hsc pm NotFOI (Just source)
             -- Bang pattern is important to avoid leaking 'tmr'
-            let !res = extractHiFileResult tmr
-            return (diags <> diags', res)
+            (diags'', !res) <- liftIO $ extractHiFileResult hsc objNeeded tmr
+            return (diags <> diags' <> diags'', res)
 
-extractHiFileResult :: Maybe TcModuleResult -> Maybe HiFileResult
-extractHiFileResult Nothing = Nothing
-extractHiFileResult (Just tmr) =
-    -- Bang patterns are important to force the inner fields
-    Just $! tmr_hiFileResult tmr
+extractHiFileResult :: HscEnv -> Bool -> Maybe TcModuleResult -> IO (IdeResult HiFileResult)
+extractHiFileResult _hsc _obj Nothing = pure ([],Nothing)
+extractHiFileResult _hsc False (Just tmr) = pure ([], Just $! HiFileResult (tmrModSummary tmr) (tmrModInfo tmr))
+extractHiFileResult hsc True (Just tmr) = do
+  (diags, linkable) <- generateObjectCode hsc tmr
+  let hmi = (tmrModInfo tmr) { hm_linkable = linkable }
+  pure (diags, Just $! HiFileResult (tmrModSummary tmr) hmi)
 
 getClientSettingsRule :: Rules ()
 getClientSettingsRule = defineEarlyCutOffNoFile $ \GetClientSettings -> do
   alwaysRerun
   settings <- clientSettings <$> getIdeConfiguration
   return (BS.pack . show . hash $ settings, settings)
 
+needsObjectCodeRule :: Rules ()
+needsObjectCodeRule = defineEarlyCutoff $ \NeedsObjectCode file -> do
+  (ms,_) <- useWithStale_ GetModSummaryWithoutTimestamps file
+  -- A file needs object code if it uses TH or any file that depends on it uses TH
+  res <-
+    if uses_th_qq ms
+    then pure True
+    else do
+      revs <- reverseDependencies file <$> useNoFile_ GetModuleGraph
+      deps <- uses_ GetModSummaryWithoutTimestamps revs
+      pure $ any uses_th_qq deps
+  pure (Just $ BS.pack $ show $ hash res, ([], Just res))
```

Comment:
`reverseDependencies` returns the transitive reverse deps.  What's the average asymptotic complexity of this rule?

Is it possible to define it recursively using the non-transitive rev deps?

Reply:
Yes, good idea!